### PR TITLE
SpaceAnalyzer: Do not reset the path to '/' after a refresh 

### PR DESCRIPTION
**Pull request by kuzux**
This needed quite a bit of editing before being able to fix the problem. By the way, the code in SpaceAnalyzer was (and still is) littered with inefficient resource usage, unsafe code and a general lack of error handling. It requires more work to get up to the standards of the rest of the codebase

### CODE REVIEW

**Review by linusg — CHANGES_REQUESTED** [review]
One commit says "Space Analyzer"

File: Userland/Applications/SpaceAnalyzer/main.cpp [hunk 1]
```diff
@@ -416,7 +379,7 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
                 continue;
             }
 
-            const SpaceAnalyzer::TreeMapNode* node = treemapwidget.path_node(k);
+            const TreeNode* node = treemapwidget.path_node(k);
```
Comment (AtkinsSJ) [inline Userland/Applications/SpaceAnalyzer/main.cpp:274]: Nit-pick: Can you move the `const` to the right side while you're editing this line?

File: Userland/Applications/SpaceAnalyzer/CMakeLists.txt [hunk 2]
```diff
@@ -7,6 +7,7 @@ compile_gml(SpaceAnalyzer.gml SpaceAnalyzerGML.h space_analyzer_gml)
 
 set(SOURCES
     TreeMapWidget.cpp
+    Tree.cpp
```
Comment (AtkinsSJ) [inline Userland/Applications/SpaceAnalyzer/CMakeLists.txt:10]: Commit message typos: "TrreeNode", and "structurees"

File: Userland/Applications/SpaceAnalyzer/Tree.cpp [hunk 3]
```diff
@@ -5,7 +5,31 @@
  */
 
 #include "Tree.h"
+#include <AK/Function.h>
```
Comment (AtkinsSJ) [inline Userland/Applications/SpaceAnalyzer/Tree.cpp:8]: Commit message typo: "priivate"

File: Userland/Applications/SpaceAnalyzer/TreeMapWidget.cpp [hunk 4]
```diff
@@ -6,7 +6,9 @@
  */
 
 #include "TreeMapWidget.h"
+#include "Tree.h"
```
Comment (AtkinsSJ) [inline Userland/Applications/SpaceAnalyzer/TreeMapWidget.cpp:9]: Commit message: "kep" looks like a typo, I think it's supposed to be "kept as"?

File: Userland/Applications/SpaceAnalyzer/main.cpp [hunk 5]
```diff
@@ -183,7 +183,10 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
 
     auto& file_menu = window->add_menu("&File");
     file_menu.add_action(GUI::Action::create("&Analyze", [&](auto&) {
-        if (auto result = analyze(tree, treemapwidget, statusbar); result.is_error()) {
+        // FIXME: Just modify the tree in memory instead of traversing the entire file system
+        // FIXME: Dispose of the old tree
```
Comment (AtkinsSJ) [inline Userland/Applications/SpaceAnalyzer/main.cpp:187]: Isn't the old tree automatically disposed of when `TreeMapWidget::set_tree()` (inside `analyze()`) assigns the new value to `m_tree`?

File: Userland/Applications/SpaceAnalyzer/main.cpp [hunk 6]
```diff
@@ -237,9 +240,9 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
             }
         }
 
-        // TODO: Refreshing data always causes resetting the viewport back to "/".
-        // It would be great if we found a way to preserve viewport across refreshes.
-        if (auto result = analyze(tree, treemapwidget, statusbar); result.is_error()) {
+        // FIXME: Dispose of the old tree
```
Comment (AtkinsSJ) [inline Userland/Applications/SpaceAnalyzer/main.cpp:243]: Same as above.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
I only have nit-picks. This is a nice improvement to some hairy code. :^)

It's making me want to make more suggestions for improvements but that's not what code review is for. :sweat_smile:

File: Userland/Applications/SpaceAnalyzer/main.cpp (hunk 5 above)
Comment (kuzux) [reply Userland/Applications/SpaceAnalyzer/main.cpp:187]: Well. the reference to the old tree inside the tree map widget is gone, but the original `RefPtr` inside the  `serenity_main` function is never changed/moved/cleaned up

File: Userland/Applications/SpaceAnalyzer/main.cpp (hunk 6 above)
Comment (kuzux) [reply Userland/Applications/SpaceAnalyzer/main.cpp:243]: Same here as well

File: Userland/Applications/SpaceAnalyzer/main.cpp (hunk 5 above)
Comment (AtkinsSJ) [reply Userland/Applications/SpaceAnalyzer/main.cpp:187]: Ah, yes. I missed that. Eventually we can stop keeping that RefPtr around I think, but one step at a time.

**Review by AtkinsSJ — APPROVED** [review]

**Pull request merged**